### PR TITLE
Make mysql cron adjustable from settings

### PR DESCRIPTION
**Pull request by geelongmicrosoldering**
# Description

After switching to mysql(mariadb) I noticed frequently while adding alias, that the database was locked. It looks like this is caused by a cron job refreshing alias every 30 minutes if a refresh is required (I'm playing with the aliases, so it is). My database is reasonably sized (200k messages?) and runs on limited hardware (Raspberry PI 3).

I considered just adjusting the cron to not refresh as often, but I decided to just add the cron job interval settings into the Settings Template instead.

Note: Probably a better way to do this. Also considered using prettyCron to display human readable translation of cron, but decided to just link cron doc instead and label it advanced

cron_settings

Fixes # cron job interval cannot be adjusted easily by admins

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added to my current server instance

### CODE REVIEW

**Comment by geelongmicrosoldering** [comment]
side note, why does this cron exist for mysql only? something i'm missing?

File: server/config/default.json
Comment (davidmckenzie) [inline server/config/default.json:15]: any reason to change the default from twice an hour?

**Comment by davidmckenzie** [comment]
> side note, why does this cron exist for mysql only? something i'm missing?

It was a workaround for the poor alias refresh time that only seems to become a real problem when you get to the scale where you need mysql anyway, iirc. I don't think there was ever a reason _not_ to use it on sqlite from my vague memory, just that there wasn't a driver for it at the time.

File: server/config/default.json
Comment (marshyonline) [reply server/config/default.json:15]: Id also like this left at twice an hour by default

**Review by marshyonline — CHANGES_REQUESTED** [review]
LGTM once default is changed back

File: server/config/default.json
Comment (geelongmicrosoldering) [reply server/config/default.json]: ah, that was a copy and paste error. default should be "0 5,35 * * * *", i believe i fixed it

**Comment by geelongmicrosoldering** [comment]
> 
> It was a workaround for the poor alias refresh time that only seems to become a real problem when you get to the scale where you need mysql anyway, iirc. I don't think there was ever a reason _not_ to use it on sqlite from my vague memory, just that there wasn't a driver for it at the time.

> 
> MySQL Blocks updates to the capcode table while the refresh is running, preventing additions of capcodes. SQLite locks out the entire UI.
> 
> There is a few PR's and Issues with the thought processes in them i think.

Makes sense. I assumed it was there for a reason. And yeah sqlite locking out the UI was the primary reason i moved to mariadb. well, that and the large db size

**Comment by geelongmicrosoldering** [comment]
@marshyonline sorry github is actually fairly foreign to me i don't fully understand the commit/review process yet. i just realised i can request a review of the change i made to fix the default cron back to twice per hour

File: server/app.js
Comment (davidmckenzie) [inline server/app.js:202]: Sorry one other change - if the config item is not present (like in upgraded instances) this will fail, needs logic to fall back on a default

**Review by davidmckenzie — CHANGES_REQUESTED** [review]

**Review by davidmckenzie — APPROVED** [review]
lgtm, thanks!

**Comment by geelongmicrosoldering** [comment]
> lgtm, thanks!

wonderful!

I still don't understand github. PR's are fairly new to me. I feel like it wants me to click something somewhere..

**Comment by davidmckenzie** [comment]
> I still don't understand github. PR's are fairly new to me. I feel like it wants me to click something somewhere..

it's just waiting on marshy's review then we can merge :)

**Comment by geelongmicrosoldering** [comment]
> > I still don't understand github. PR's are fairly new to me. I feel like it wants me to click something somewhere..
> 
> it's just waiting on marshy's review then we can merge :)

ah, awesome that's cool I just didn't know if i did it right haha

**Review by DanrwAU — CHANGES_REQUESTED** [review]
The only thing I can see is that it needs error handling for invalid values on the settings page as if you put gibberish in, it crashes the application on startup. 
Not sure which is easier, i'd almost be leaning towards GUI validation rather than a check serverside. 

Sorry to bounce this one back again, other than that it looks good.

**Comment by DanrwAU** [comment]
@geelongmicrosoldering 

I fixed this one up, Added a node module to validate cron strings in app.js.

I also removed the "required" from the settings field as it would prevent users saving other settings until the default is pasted in.

**Comment by geelongmicrosoldering** [comment]
> @geelongmicrosoldering
> 
> I fixed this one up, Added a node module to validate cron strings in app.js.
> 
> I also removed the "required" from the settings field as it would prevent users saving other settings until the default is pasted in.

Hey, thanks! I'm on my phone but looks good!

**Review by DanrwAU — DISMISSED** [review]

**Comment by DanrwAU** [comment]
@marshyonline @davidmckenzie if one of you 2 could look this one over. since i've now monkeyed with the code 🤣

**Comment by geelongmicrosoldering** [comment]
> @marshyonline @davidmckenzie if one of you 2 could look this one over. since i've now monkeyed with the code 🤣

Actually worth querying, does null/empty/undefined validate with that validator?

If not you can probably remove my fallback line above because validating will revert the falsy value to default in your validation check anyway.

**Comment by geelongmicrosoldering** [comment]
@DanrwAU Ah looks like that won't validate

cron.isValidCron('* * * * * *', { seconds: true });

If you pass Cron with seconds but don't tell the validator there's seconds it returns false according to docs

**Comment by DanrwAU** [comment]
> @DanrwAU Ah looks like that won't validate
> 
> cron.isValidCron('* * * * * *', { seconds: true });
> 
> If you pass Cron with seconds but don't tell the validator there's seconds it returns false according to docs

Anyone giving it settings deserves it to fail. 

🤣  i'll leave it up to the group, but for this usecase i can't think of a single reason anyone would ever want to use seconds. Infact it may actually be detrimental as you could end up in a loop where it is running continuously if you set it too short?

**Comment by geelongmicrosoldering** [comment]
> > @DanrwAU Ah looks like that won't validate
> > cron.isValidCron('* * * * * *', { seconds: true });
> > If you pass Cron with seconds but don't tell the validator there's seconds it returns false according to docs
> 
> Anyone giving it settings deserves it to fail.
> 
>  i'll leave it up to the group, but for this usecase i can't think of a single reason anyone would ever want to use seconds. Infact it may actually be detrimental as you could end up in a loop where it is running continuously if you set it too short?

The default has seconds lol, that was my point haha. "0 5,35 * * * *" is 6 parameters, we were already passing seconds

Translates to every 5m0s, 35m0s. Thats why it starts with 0, otherwise it would occur every second on the 5th and 35th minute

**Comment by DanrwAU** [comment]
> > > @DanrwAU Ah looks like that won't validate
> > > cron.isValidCron('* * * * * *', { seconds: true });
> > > If you pass Cron with seconds but don't tell the validator there's seconds it returns false according to docs
> > 
> > 
> > Anyone giving it settings deserves it to fail.
> > i'll leave it up to the group, but for this usecase i can't think of a single reason anyone would ever want to use seconds. Infact it may actually be detrimental as you could end up in a loop where it is running continuously if you set it too short?
> 
> The default has seconds lol, that was my point haha. "0 5,35 * * * *" is 6 parameters, we were already passing seconds
> 
> Translates to every 5m0s, 35m0s. Thats why it starts with 0, otherwise it would occur every second on the 5th and 35th minute

Interesting.... it worked when i tested it 😕

I'll take a look when i get a sec

**Comment by davidmckenzie** [comment]
fwiw, here's a thread re validation on the particular module we use: https://github.com/kelektiv/node-cron/issues/419

**Comment by geelongmicrosoldering** [comment]
Sorry guys, I've had heaps on and totally temporarily abandoned this 8 months ago.

What was pending that i needed to fix?

@marshyonline @davidmckenzie

**Comment by geelongmicrosoldering** [comment]
FWIW, can confirm that adding ` { seconds: true }` to cron validator, the behaviour is now as expected. before then, the custom cron was ignored because it didn't validate, so the default 5,35 was used instead.

Server now happy

**Comment by marshyonline** [comment]
Ping for re-review plz

**Review by marshyonline — APPROVED** [review]
Looks ok to me

**Pull request merged**
